rfcs/20190305-modular-tensorflow.md

```diff
@@ -0,0 +1,481 @@
+# Modular TensorFlow
+
+| Status        | Proposed                                             |
+:-------------- |:---------------------------------------------------- |
+| **Author(s)** | Gunhan Gulsoy (gunan@google.com)                     |
+| **Sponsor**   | Martin Wicke (wicke@google.com)                      |
+| **Updated**   | 2019-03-06                                           |
+
+
+## Motivation
+
+TensorFlow is a very successful open source project. Since it has been open sourced, [1800+ contributors](https://github.com/tensorflow/tensorflow) have submitted code into TF from outside Google. However, as more and more developers contribute, it becomes more and more difficult to manage contributions in the single repository.
+
+This project aims to split the TensorFlow codebase into **smaller, more focused**, repositories that can be released and managed separately. These modules will talk to each other using **well defined APIs** that guarantee backwards compatibility. Thanks to the module APIs, these modules are now **managed/owned/released independently.**
+
+### Problems addressed
+
+#### Spaghetti dependencies
+
+Everything being in a single repository encourages everyone to freely use all other code in the repository, even when this does not make sense. This ensures the library is very difficult to modify later.
+
+#### Long build times
+
+Many volunteers and developers outside Google use their laptops for development. On such systems, TF development cycles require building all of tensorflow that takes around 2 hours. While there is an argument to be made that bazel caching should ensure that they build all of TF only once, without forge, bazel caching is not working as [well as expected](https://docs.bazel.build/versions/master/remote-caching.html#known-issues).
+
+#### Adding support for new hardware is very difficult and not scalable
+
+The ML ecosystem is constantly expanding. New hardware for accelerating ML applications is being worked on by many teams inside and outside Google. As the most popular machine learning framework, TF is expected to add support for many of these hardware as quickly as possible.
+
+Currently, this means that all such hardware developers need to check in their code into the [main tensorflow repository](http://github.com/tensorflow/tensorflow). This means that all the changes are required to go through TF team's review. This can make merging support for new hardware very very difficult.
```

Comment:
... or maintain a fork of TF, right?

Reply:
The kernel way without GPL enforcing ends you up with Caffe, as the saying goes by, “there are a thousand TensorFlow in a thousand contributors’ eyes”.

Reply:
Yes, here are two examples I am aware of:
https://github.com/ROCmSoftwarePlatform/tensorflow-upstream
https://github.com/codeplaysoftware/tensorflow